dali/pipeline/util/bounding_box_utils.h

```diff
@@ -0,0 +1,243 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.c++ copyswa
+
+#ifndef DALI_PIPELINE_UTIL_BOUNDING_BOX_UTILS_H_
+#define DALI_PIPELINE_UTIL_BOUNDING_BOX_UTILS_H_
+
+#include <limits>
+#include "dali/core/geom/box.h"
+#include "dali/core/tensor_layout.h"
+#include "dali/core/math_util.h"
+
+namespace dali {
+
+template <int ndim>
+void AxisFlip(Box<ndim, float>& box, int axis) {
+  assert(axis >= 0 && axis < ndim);
+  box.lo[axis] = 1.0f - box.hi[axis];
+  box.hi[axis] = 1.0f - box.lo[axis];
+}
+
+template <int ndim>
+void HorizontalFlip(Box<ndim, float>& box) {
+  static_assert(ndim >= 1);
+  AxisFlip(box, 0);
+}
+
+template <int ndim>
+void VerticalFlip(Box<ndim, float>& box) {
+  static_assert(ndim >= 2);
+  AxisFlip(box, 1);
+}
+
+template <int ndim>
+void DepthWiseFlip(Box<ndim, float>& box) {
+  static_assert(ndim >= 3);
+  AxisFlip(box, 2);
+}
+
+template <int ndim>
+constexpr TensorLayout DefaultLayout() {
+  assert(ndim == 2 || ndim == 3);
+  if (ndim == 3)
+    return TensorLayout{"xyzXYZ"};
+  else
+    return TensorLayout{"xyXY"};
+}
+
+template <int ndim>
+constexpr TensorLayout DefaultAnchorAndShapeLayout() {
+  assert(ndim == 2 || ndim == 3);
+  if (ndim == 3)
+    return TensorLayout{"xyzWHD"};
+  else
+    return TensorLayout{"xyWH"};
+}
+
+template <int ndim>
+float AspectRatio(const Box<ndim, float>& box, int dim0, int dim1) {
+  assert(dim0 >= 0 && dim0 < ndim);
+  assert(dim1 >= 0 && dim1 < ndim);
+  auto extent = box.extent();
+  auto extent0 = extent[dim0];
+  auto extent1 = extent[dim1];
+  return extent0 / extent1;
+}
+
+template <int ndim>
+float MinAspectRatio(const Box<ndim, float>& box) {
+  float min_ar = std::numeric_limits<float>::max();
+  for (int i = 0; i < ndim; i++) {
+    for (int j = i + 1; j < ndim; j++) {
+      min_ar = std::min(min_ar, AspectRatio(box, i, j));
+    }
+  }
+  return min_ar;
+}
+
+template <int ndim>
+float MaxAspectRatio(const Box<ndim, float>& box) {
+  float max_ar = std::numeric_limits<float>::min();
+  for (int i = 0; i < ndim; i++) {
+    for (int j = i + 1; j < ndim; j++) {
+      max_ar = std::max(max_ar, AspectRatio(box, i, j));
+    }
+  }
+  return max_ar;
+}
+
+template <int ndim>
+Box<ndim, float> Uniform(float min, float max) {
+  return Box<ndim, float>(vec<ndim, float>(min), vec<ndim, float>(max));
+}
+
+template <typename Coords>
+void PermuteCoords(Coords& coords,
+                   TensorLayout orig_layout,
+                   TensorLayout new_layout) {
+  if (orig_layout == new_layout)
+    return;
+  assert(orig_layout.is_permutation_of(new_layout));
+  assert(coords.size() == orig_layout.size());
+  auto perm = GetDimIndices(orig_layout, new_layout);
+  auto out = coords;
+  for (int d = 0; d < static_cast<int>(coords.size()); d++)
+    out[d] = coords[perm[d]];
+  std::swap(coords, out);
+}
+
+template <int ndim>
+void ReadBoxes(span<Box<ndim, float>> boxes,
```

Comment:
Can I ask for some docstrings for Read and WriteBoxes?

Where they are read from and where to?

Reply:
sure